### PR TITLE
Annotate results that are used within the same file

### LINKED ISSUE #38: Is it possible to exclude results that are used within the same file?
I have a lot of what I would consider to be false positives. A method is not being used as in import anywhere, but it being used within the same file. Is there a way to filter these out?

### PR DESCRIPTION
**Pull request by danvk**
Fixes #38 

With this PR, exports that are never imported but are still used within their own module are annotated as such:

```
$ ts-prune
src/B.ts:9 - UnusedFooType
src/C.ts:9 - default
src/cities.ts:1 - sepehub
src/cities.ts:2 - kuariob
src/cities.ts:4 - femvacsah
src/cities.ts:5 - sijelup
src/internal-uses.ts:5 - usedInThisFile (used in module)
src/internal-uses.ts:7 - thisOneIsUnused
src/internal-uses.ts:9 - UsedInThisFile (used in module)
src/internal-uses.ts:11 - Unused
src/internal-uses.ts:13 - Row (used in module)
src/internal-uses.ts:17 - UnusedProps
src/dynamic/fail.ts:1 - foo
src/dynamic/fail.ts:3 - bar
```

If you don't consider these symbols problems, you can grep them out with:

```
ts-prune | grep -v 'used in module'
```

If you do consider these problems, you can ignore the "(used in module") or use `sed` or some such to get rid of it.

### CODE REVIEW

**Comment by nadeesha** [comment]
I think this is a worthwhile addition. Thanks @danvk.

File: src/analyzer.ts [hunk 1]
```diff
@@ -117,6 +120,15 @@ const emitDefinitelyUsed = (file: SourceFile, onResult: OnResultType) => {
 const emitPotentiallyUnused = (file: SourceFile, onResult: OnResultType) => {
   const exported = getExported(file);
 
+  const idsInFile = file.getDescendantsOfKind(ts.SyntaxKind.Identifier);
+  const referencedInFile = _(idsInFile)
```
Comment (nadeesha) [inline src/analyzer.ts:124]: Would you kindly refactor this to not use `_` chaining please? You're free to use the `fp` constructs as you wish :)

Somethings like:

```ts
const referenceCounts = countBy((idsInFile || []).map(node => node.getText()));
const referencedInFile = Object.entries(referenceCounts)...
```

Comment (danvk) [reply src/analyzer.ts]: Sure, done, though I am curious what the objection to chaining is. It has to be `countBy(x => x)(...)`, the iteratee seems to be required w/ lodash-fp.

**Comment by danvk** [comment]
Thanks @nadeesha. I was thinking that one way to merge this while keeping the spirit of "zero configuration" would be to add some sort of marker to the output for internally-used exports. That way you can grep them out if you're not interested.

Perhaps something like:

```
$ ts-prune
src/internal-uses.ts:5 - usedInThisFile (exported but used in module)
src/internal-uses.ts:7 - thisOneIsUnused
src/internal-uses.ts:9 - UsedInThisFile (exported but used in module0
src/internal-uses.ts:11 - Unused
src/internal-uses.ts:17 - UnusedProps
```

Or if that's too verbose, then:

```
$ ts-prune
src/internal-uses.ts:5 - usedInThisFile *
src/internal-uses.ts:7 - thisOneIsUnused
src/internal-uses.ts:9 - UsedInThisFile *
src/internal-uses.ts:11 - Unused
src/internal-uses.ts:17 - UnusedProps
```

Let me know what you think.

**Comment by danvk** [comment]
Hi @nadeesha, I've gone ahead and implemented the idea of tweaking the output to reflect whether symbols are used in their own modules:

```
src/B.ts:9 - UnusedFooType
src/C.ts:9 - default
src/cities.ts:1 - sepehub
src/cities.ts:2 - kuariob
src/cities.ts:4 - femvacsah
src/cities.ts:5 - sijelup
src/internal-uses.ts:5 - usedInThisFile (used in module)
src/internal-uses.ts:7 - thisOneIsUnused
src/internal-uses.ts:9 - UsedInThisFile (used in module)
src/internal-uses.ts:11 - Unused
src/internal-uses.ts:13 - Row (used in module)
src/internal-uses.ts:17 - UnusedProps
src/dynamic/fail.ts:1 - foo
src/dynamic/fail.ts:3 - bar
```

If you don't consider these symbols problems, you can grep them out with:

```
ts-prune | grep -v 'used in module'
```

We can tweak the wording / formatting, of course. Let me know what you think.

File: src/analyzer.ts (hunk 1 above)
Comment (nadeesha) [reply src/analyzer.ts]: You can write it alternatively as `countBy(x => x, input)` as well.

It's more of a preference to use the `fp` methods, than a hesitation to use chaining. I find the `fp` methods work on immutable data, have consistent interfaces, and can be chained with `pipe`.

**Pull request merged**